File: packages/api/src/coding-schedule.ts
/**
 * Phase 68-C7 — Coding schedules (Paseo schedule analogue).
 *
 * A schedule creates a **new** Coding workspace on cron for a fixed project
 * cwd + harness, then runs the prompt once.
 *
 * Distinct from heartbeats (wake existing workspace). Not Team jobs / Chains.
 * Not EnvoyAI TriggerStore digests.
 */

import {
  CODING_HEARTBEAT_CRON_PRESETS,
  isValidCodingHeartbeatCron,
  type CodingHeartbeatCronPreset,
} from "./coding-heartbeat.js"
import { isCodingHarnessId, type CodingHarnessId } from "./coding-harness.js"

export type CodingScheduleHarness = CodingHarnessId

export type CodingSchedule = {
  id: string
  name: string
  /** 5-field cron (minute hour day-of-month month day-of-week). */
  cron: string
  /**
   * Optional IANA timezone. MVP ticker evaluates cron in **UTC**
   * (same as `isCronMatch`); timezone is stored for forward-compat / UI hint.
   */
  timezone?: string
  prompt: string
  /** Project folder (absolute path on the home node). */
  cwd: string
  harness: CodingScheduleHarness
  enabled: boolean
  maxRuns?: number
  runCount: number
  lastFiredAt?: string
  lastError?: string
  /** chatId / sessionId created on last fire. */
  lastWorkspaceId?: string
  createdAt: string
  updatedAt: string
}

export type CreateCodingScheduleInput = {
  name: string
  cron: string
  timezone?: string
  prompt: string
  cwd: string
  harness: CodingScheduleHarness
  enabled?: boolean
  maxRuns?: number
}

export type UpdateCodingScheduleInput = {
  id: string
  enabled?: boolean
  cron?: string
  prompt?: string
  name?: string
  cwd?: string
  harness?: CodingScheduleHarness
}

/** Soft cap — keep the ticker honest and the list UI small. */
export const MAX_CODING_SCHEDULES = 20

/** Shared cron presets (UTC) — same as heartbeats. */
export const CODING_CRON_PRESETS = CODING_HEARTBEAT_CRON_PRESETS

export type CodingCronPreset = CodingHeartbeatCronPreset

export function isCodingScheduleHarness(
  value: unknown,
): value is CodingScheduleHarness {
  return typeof value === "string" && isCodingHarnessId(value)
}

/** Basic 5-field cron shape check (does not expand steps). */
export function isValidCodingScheduleCron(cron: string): boolean {
  return isValidCodingHeartbeatCron(cron)
}

File: apps/envoygo/lib/coding/coding_schedule.dart
/// Phase 68-C7 — Coding schedules (Dart port of packages/api coding-schedule).
library;

import 'coding_heartbeat.dart';

const maxCodingSchedules = 20;

class CodingSchedule {
  const CodingSchedule({
    required this.id,
    required this.name,
    required this.cron,
    required this.prompt,
    required this.cwd,
    required this.harness,
    required this.enabled,
    required this.runCount,
    required this.createdAt,
    required this.updatedAt,
    this.timezone,
    this.maxRuns,
    this.lastFiredAt,
    this.lastError,
    this.lastTaskId,
  });

  final String id;
  final String name;
  final String cron;
  final String? timezone;
  final String prompt;
  final String cwd;
  final String harness;
  final bool enabled;
  final int? maxRuns;
  final int runCount;
  final String? lastFiredAt;
  final String? lastError;
  final String? lastTaskId;
  final String createdAt;
  final String updatedAt;

  factory CodingSchedule.fromJson(Map<String, dynamic> json) {
    return CodingSchedule(
      id: (json['id'] as String?)?.trim() ?? '',
      name: (json['name'] as String?)?.trim() ?? '',
      cron: (json['cron'] as String?)?.trim() ?? '',
      timezone: (json['timezone'] as String?)?.trim(),
      prompt: (json['prompt'] as String?)?.trim() ?? '',
      cwd: (json['cwd'] as String?)?.trim() ?? '',
      harness: (json['harness'] as String?)?.trim() ?? 'envoy-harness',
      enabled: json['enabled'] != false,
      maxRuns: json['maxRuns'] is int ? json['maxRuns'] as int : null,
      runCount: json['runCount'] is int ? json['runCount'] as int : 0,
      lastFiredAt: (json['lastFiredAt'] as String?)?.trim(),
      lastError: (json['lastError'] as String?)?.trim(),
      lastTaskId: (json['lastTaskId'] as String?)?.trim(),
      createdAt: (json['createdAt'] as String?)?.trim() ?? '',
      updatedAt: (json['updatedAt'] as String?)?.trim() ?? '',
    );
  }

  Map<String, dynamic> toCreateJson() => {
        'name': name,
        'cron': cron,
        if (timezone != null && timezone!.isNotEmpty) 'timezone': timezone,
        'prompt': prompt,
        'cwd': cwd,
        'harness': harness,
        'enabled': enabled,
        if (maxRuns != null) 'maxRuns': maxRuns,
      };
}

class CreateCodingScheduleInput {
  const CreateCodingScheduleInput({
    required this.name,
    required this.cron,
    required this.prompt,
    required this.cwd,
    required this.harness,
    this.timezone,
    this.enabled = true,
    this.maxRuns,
  });

  final String name;
  final String cron;
  final String? timezone;
  final String prompt;
  final String cwd;
  final String harness;
  final bool enabled;
  final int? maxRuns;

  Map<String, dynamic> toJson() => {
        'name': name.trim(),
        'cron': cron.trim(),
        if (timezone != null && timezone!.trim().isNotEmpty)
          'timezone': timezone!.trim(),
        'prompt': prompt.trim(),
        'cwd': cwd.trim(),
        'harness': harness.trim(),
        'enabled': enabled,
        if (maxRuns != null) 'maxRuns': maxRuns,
      };
}

bool isValidCodingScheduleCron(String cron) =>
    isValidCodingHeartbeatCron(cron);
